[Woo POS] Move `Analytics` protocols to WooFoundation framework

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Depends on https://github.com/woocommerce/woocommerce-ios/pull/12677
<!-- Id number of the GitHub issue this PR addresses. -->

## Why

The goal of this PR is to move the `Analytics` protocol from the app framework to the WooFoundation framework so that the analytics instance can be DI'ed to the new POS framework. The usage of this change will be in a separate PR, with an example in https://github.com/woocommerce/woocommerce-ios/commit/95a8554e90f5cc543b5ef9af60316d6747877998 where a randomly picked event is tracked when the POS view appears.

## How

Some decisions that I made for the initial implementation, any early feedback is welcome and I can make any changes before making the PR official for review!

### What's moved to WooFoundation

- `Analytics` protocol
- Dependencies in the protocol
  - `AnalyticsProvider` protocol 
  - `WooAnalyticsEventPropertyType` protocol 
  - `WooAnalyticsStat` enum:
    - This contains the most diffs (1231 new/deleted lines)
    - I was debating about whether to move this, or updating its usage in `Analytics` to be a more generic string enum. In the end, I decided to include this because:
      - It will require quite some changes to update the enum usage in the protocol
      - It's nice having a consolidated list of events to avoid accidentally having two events of the same name
  - `WooAnalyticsEvent` struct:
    - This one isn't necessary, since this struct is to allow type-checking on the event properties and a cleaner way of specifying events. But I thought this could make event tracking code easier in the POS framework as well

### Why WooFoundation

As WooFoundation can be the lowest level of dependency for any framework with UI, this feels like a good place to have the **protocols** (no implementation) of major dependencies. This way, any framework (like POS) that depends on it can specify the dependency in protocol, and then the framework caller can pass whatever implementation to the framework object.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

- [x] @jaclync tests that Tracks events tracked in the app still show up in Tracks

Feel free to do a confidence test to ensure the Tracks events from app usage still show up in the console and Tracks.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

<img width="527" alt="Screenshot 2024-05-09 at 9 11 29 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/f831ad85-2783-43a7-b659-6aac82d3d506">


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync committed:
Add Analytics protocol with related enum/struct/protocol's to WooFoundation.

jaclync committed:
Replace analytics dependencies in the app framework with WooFoundation.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

jaclync committed:
Fix import build errors in unit tests.

bozidarsevo commented:
Do we need to update https://github.com/woocommerce/woocommerce-ios/blob/trunk/docs/tracking-events.md?
Maybe just to write that `import protocol WooFoundation.Analytics` needs to be used?

jaclync committed:
Add import reference to the tracking-events doc.

jaclync committed:
Reorder imports in alphabetical order.

jaclync commented:
> Do we need to update https://github.com/woocommerce/woocommerce-ios/blob/trunk/docs/tracking-events.md?
> Maybe just to write that `import protocol WooFoundation.Analytics` needs to be used?

Yup good idea! I added some notes in the last 2 commits.

iamgabrielma commented on WooCommerce/Classes/Analytics/AppStartupWaitingTimeTracker.swift:
```diff
@@ -1,5 +1,6 @@
 import Foundation
 import Yosemite
+import WooFoundation
```
Just for consistency, we can change this one to:
```suggestion
import protocol WooFoundation.Analytics
```

jaclync replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/12685/commits/725cf2a14bc1b3633b2a10caab220e28061dee49

iamgabrielma commented on WooCommerce/Classes/Analytics/WooAnalytics.swift:
```diff
@@ -5,22 +5,26 @@ import WordPressShared
 import WidgetKit
 import enum Alamofire.AFError
 import Yosemite
+import enum WooFoundation.WooAnalyticsStat
+import protocol WooFoundation.Analytics
+import protocol WooFoundation.AnalyticsProvider
+import struct WooFoundation.WooAnalyticsEvent
```
The type import here make me wonder if `WooAnalyticsStat` and `WooAnalyticsEvent` should be in WooFoundation, or be kept in WooCommerce now that we've moved the protocols over:

While `WooAnalyticsEvent` is nice for consistency for event tracking, it feels that `WooAnalyticsStat` now becomes an implementation detail of the WooCommerce app, and since what we want to track in the POS framework (and potentially separate app) would most likely be be different events, `WooAnalyticsStat` perhaps should stay in WooCommerce. Thoughts?

jaclync replied:
Yup I had similar thoughts before and these aren't necessary indeed, I shared the reasons in the description with the following highlights - 

- `WooAnalyticsStat` enum: It's nice having a consolidated list of events to avoid accidentally having two events of the same name
- `WooAnalyticsEvent` struct: this one is an extension of the protocol and isn't necessary, since this struct is to allow type-checking on the event properties and a cleaner way of specifying events. But I thought this could make event tracking code easier in the POS framework as well

WDYT + @joshheald @bozidarsevo? I'm happy to make any changes after we make a decision.

joshheald replied:
>  It's nice having a consolidated list of events to avoid accidentally having two events of the same name.

This makes sense, but at the same time, the existing events are (as @iamgabrielma says) closely coupled to the existing app. We're likely to _want_ to have two events with the same name, e.g. `collect_payment_tapped`

Perhaps we should discuss with the wider team the idea of having a different prefix for POS events instead, i.e. `woocommerceposios_collect_payment_tapped` to namespace these separately. If we did that, we can leave the enum and struct in the existing app and just extract a protocol for `WooAnalyticsEvent` to put in `WooFoundation`.

We should do the same thing on both platforms though.

joshheald replied:
Also... if we do need to import 4 things from a framework, it feels to me like it's time to start importing the framework. Just my 2¢, but individual lines for each file adds clutter without value.

jaclync replied:
Thanks for the feedback @iamgabrielma @joshheald! It really helped clear up my original doubts and debates. I've moved the event enum and struct back to the main app in https://github.com/woocommerce/woocommerce-ios/pull/12685/commits/264b8b48666682a122f63880030661a8165a4a71 with a bunch of other updates and polishes in the latest commits.

> Perhaps we should discuss with the wider team the idea of having a different prefix for POS events instead, i.e. `woocommerceposios_collect_payment_tapped` to namespace these separately. If we did that, we can leave the enum and struct in the existing app and just extract a protocol for `WooAnalyticsEvent` to put in `WooFoundation`.
> 
> We should do the same thing on both platforms though.

If we have plans to move POS to a separate app, personally I think keeping the `woocommerceios_` prefix with the additional POS mention like `woocommerceios_pos_` would be easier to differentiate events between the 2 apps. In any case, for the context in this PR, no decision is made in the POS framework and we can defer the POS implementation later.

iamgabrielma commented on WooFoundation/WooFoundation/Protocols/Analytics/WooAnalyticsStat.swift:
```diff
@@ -0,0 +1,1231 @@
+/// This enum contains all of the events we track in the app. Please reference the "Woo Mobile Events Draft i2"
+/// spreadsheet for more details.
+///
+/// One of goals of this `enum` is to be able to list all the event names that we use throughout
+/// the app. We can also potentially make a parser to gather all the event names and automatically
+/// compare them with WCAndroid. With that, we can make sure both platforms are tracking the
+/// same events. Right now, we use the spreadsheet. XD
```
We can update some of these comments:
```suggestion
/// This enum contains all of the events we track in the WooCommerce app.
///
/// One of goals of this `enum` is to be able to list all the event names that we use throughout
/// the app.
```

jaclync replied:
Nice catch! Updated in https://github.com/woocommerce/woocommerce-ios/pull/12685/commits/725cf2a14bc1b3633b2a10caab220e28061dee49

iamgabrielma reviewed (approved):
Thanks for working on this. Looks good and tests well! Left a nit and a question regarding some of the types that we're moving over.

joshheald commented on WooCommerce/Classes/Analytics/WooAnalytics.swift:
```diff
@@ -175,7 +179,7 @@ public extension WooAnalytics {
 extension WooAnalytics {
 
-    func setUserHasOptedOut(_ optedOut: Bool) {
+    public func setUserHasOptedOut(_ optedOut: Bool) {
```
The `public` declarations in this file seem strange, if `WooAnalytics` itself is staying in the main target. Nothing should be accessing these from outside the app framework, AFAIK.

Implementing functions of a public protocol doesn't have a requirement that the implementations be public too... they only need to be public if they're declared as public (per function) in the protocol.

I can see that WooAnalytics itself has been public since 2019... and I'm not really sure why that is. Perhaps it started life in another framework?

jaclync replied:
Yup that was a legacy decision that predates my time, and this `public` diff was because of `WooAnalytics` being public. In https://github.com/woocommerce/woocommerce-ios/pull/12685/commits/beaf2d6d0bcf07aecd5bc2b524b8407536d794fb, I updated `WooAnalytics` to internal and no build errors as expected.

joshheald commented on WooCommerce/Classes/Analytics/WooAnalyticsEvent+AppIntents.swift:
```diff
@@ -1,3 +1,5 @@
+import struct WooFoundation.WooAnalyticsEvent
```
Seeing all these imports makes me feel like `WooAnalyticsEvent` should stay in WooCommerce, and instead implement a protocol from WooFoundation (`AnalyticsEvent`). The POS can then have `WooPOSAnalyticsEvent` which implements it too.

jaclync replied:
`WooAnalyticsEvent` has been moved back to the main app as of the latest PR. As `Analytics.track(event: WooAnalyticsEvent)` is an extension of `Analytics`, I decided to not include this extension with an event protocol in this PR given the PR size. When we're ready to implement event tracking in the POS framework, we can make this protocol change as we see fit.

jaclync committed:
Move `WooAnalyticsStat` and `WooAnalyticsEvent` from WooFoundation back to app framework per PR feedback.

jaclync committed:
Remove empty lines from import removal and duplicate comments.

jaclync committed:
Update import and `WooAnalyticsStat` enum comment based on PR feedback.

jaclync committed:
Fix unit testing build errors from `Analytics` protocol conformance in `TestAnalytics`.

jaclync committed:
Update `WooAnalytics` implementation from public to internal as it's specific to the app framework.

jaclync committed:
Update comment in the tracking-events doc to remove the enum and struct reference.

jaclync committed:
Separate analytics event tracking property updates to WooAnalyticsStat specific and error specific in WooAnalytics and Analytics extensions respectively.

jaclync committed:
Remove unused functions in `Analytics` protocol.

jaclync commented on WooCommerce/Classes/Analytics/WooAnalytics.swift:
```diff
@@ -122,16 +175,31 @@ public extension WooAnalytics {
             return
         }
 
-        let properties = combinedProperties(from: error, with: passedProperties)
```
The error properties logic has been moved to the `WooAnalytics.track(_ eventName:properties:error:)` since the error properties are independent of `WooAnalyticsStat`

jaclync commented:
Update: I've read that we're pausing the framework work in pfoUAQ-EF-p2#comment-517. Please skip reviewing this PR today and I'll post an update when the decision is confirmed. The following was drafted before I heard about the news.


--- 

Thanks for all the suggestions and feedback above! I believe I've responded to all the comments and updated the PR based on them, it's ready for another pass. The diffs now are much smaller, though there are still lots of changed files from the `Analytics` protocol import. I've done a pass of the import review myself to ensure there are no empty lines etc., and I'd recommend focusing on these files:

- `Analytics` protocol: the stat enum parameter has been replaced with a string, so that the POS framework doesn't have to be tied to using a String enum. 3 tracking methods are also removed as the main app event tracking use cases are tracking with `WooAnalyticsStat`. Only `track(_ eventName:properties:error:)` is necessary with the error properties logic for reuse in the POS framework
- `WooAnalytics` file: for the app to keep using the `WooAnalyticsStat` enum in all existing event tracking use cases, the new `track(_ eventName:properties:error:)` updates the properties with an optional error and triggers remote event tracking. The original `track(_ stat:properties:error:)` updates the properties based on the `WooAnalyticsStat` enum as some cases add site properties - this is specific to the main app and the POS doesn't need to know

jaclync commented:
Update: as other POS work is based on this PR and all the changes are within WooCommerce & WooFoundation frameworks, please continue with review today @iamgabrielma @joshheald @bozidarsevo 🙇🏻‍♀️

iamgabrielma commented on WooCommerce/Classes/Analytics/WooAnalytics.swift:
```diff
@@ -5,8 +5,10 @@ import WordPressShared
+import protocol WooFoundation.AnalyticsProvider
 
-public class WooAnalytics: Analytics {
+final class WooAnalytics: Analytics {
```
Nice catch!

iamgabrielma commented:
2nd pass LGTM!

joshheald commented on WooCommerce/Classes/App Intents/CreateOrderAppIntent.swift:
```diff
@@ -11,7 +11,7 @@ struct CreateOrderAppIntent: AppIntent {
     func perform() async throws -> some IntentResult {
         MainTabBarController.presentOrderCreationFlow()
-        ServiceLocator.analytics.track(event: WooAnalyticsEvent.AppIntents.shortcutWasOpened(with: .createOrder))
+        ServiceLocator.analytics.track(event: .AppIntents.shortcutWasOpened(with: .createOrder))
```
A bit of a nitpick, but these removals of the full specifier make it a little more difficult to review, and AFAIK don't contribute to the change. Were they needed when `WooAnalyticsEvent` was in WooFoundation, earlier in this PR's life?

jaclync replied:
Yup these removals were from the previous `WooAnalyticsEvent` move to the WooFoundation framework and aren't required for the latest state of the PR. As I couldn't find a trivial way to revert these like how I can use regex replace for the imports, plus I think it makes the code simpler, I'll keep these removals 😅

joshheald reviewed (approved):
LGTM. One small nit but it's not a big deal really :) 

Thanks for the changes
